sklearn/linear_model/_base.py

```diff
@@ -287,11 +308,25 @@ def predict(self, X):
 
     def _set_intercept(self, X_offset, y_offset, X_scale):
         """Set the intercept_"""
+
+        xp, _ = get_namespace(X_offset, y_offset, X_scale)
+
         if self.fit_intercept:
             # We always want coef_.dtype=X.dtype. For instance, X.dtype can differ from
             # coef_.dtype if warm_start=True.
-            self.coef_ = np.divide(self.coef_, X_scale, dtype=X_scale.dtype)
-            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
+            coef_ = xp.astype(self.coef_, X_scale.dtype, copy=False)
+            coef_ = self.coef_ = xp.divide(coef_, X_scale)
+
+            if coef_.ndim == 1:
+                intercept_ = y_offset - X_offset @ coef_
+            else:
+                intercept_ = y_offset - X_offset @ coef_.T
+
+            if y_offset.ndim < 1:
+                intercept_ = _item(intercept_, xp=xp)
```

Comment:
I don't understand why would this `_item` helper would ever be needed:


```python
>>> import array_api_strict as xp
>>> X_offset = xp.mean(xp.ones(shape=(10, 3)), axis=0)
>>> y = xp.ones(shape=10)
>>> y_offset = xp.mean(y)
>>> y_offset.ndim
0
>>> X = xp.ones(shape=(10, 3))
>>> X_offset = xp.mean(X, axis=0)
>>> X_offset.shape
(3,)
>>> coef = xp.ones(3)
>>> intercept = y_offset - - X_offset @ coef
>>> intercept.ndim
0
```

I also tried with `array_api_compat.torch` and get the same shape for the intercept.

Reply:
If you remove it, a doctest start failing. I investigated further, it actually does something only for `numpy` inputs, it converted them from `np.array` to `np.float` types. Anyway I remove `_item` entirely and will find a better way for passing the doctest.

Reply:
So @ogrisel @betatim I managed to remove `_item` and I understand what was happening now.

Turns out that `x1=np.array(3.14)` has dtype `np.ndarray`  and `repr(x1)` is `array(3.14)` but `x2 = np.array(3.14).reshape((-1,))[0]` has dtype `np.float64` and `repr(x2)` is `3.14`. **numpy is the only library I have seen to have this inconsistency** so `_item` was indeed a noop for all other cases.

It was necessary to pass the doctest of the `ridge_regression` function, because even if `x1 == x2`, this doctestwill fail because doctests validate equality of outputs by checking that their respective `repr` are equal.

IIRC You had the same issue a while ago with some metric that started to output `np.array(somefloat)` and broke a doctest and you fixed it by forcing it to a `float` that `repr` correctly.

Here, just forcing to `float` was breaking other tests that were using `.shape` and `.dtype` attributes, that of course a normal `float` doesn't exhibit. I could have forced `float` and fix the test with an exception in this case but liked it less.

Anyway, I investigated a bit deeper why rewriting functions with array api style here turned initial `np.float64` type into `np.ndarray` type, and it turns out that there was a mistake that caused that: a `[0]` was missing [there](https://github.com/scikit-learn/scikit-learn/pull/27800/files#diff-ccd606dafadeff9924fd14b1c68b34049f400eeb30c94c507c5143ee4ccfc86bR203) (I didn't know at the time that `xp.zeros(1,...)` and `np.zeros(1,...)` return an object of shape `(1,)` and not `()`. With this fix there is no change or regression with `numpy` behavior when comparing to main.